session-state-machine/src/main/scala/zio/raft/sessionstatemachine/SessionStateMachine.scala

```diff
@@ -0,0 +1,499 @@
+      val metadata = SessionMetadata(
+        capabilities = cmd.capabilities,
+        createdAt = Instant.EPOCH  // Placeholder - will be set properly in integration
+      )
+      
+      val stateWithMetadata = state.updated["metadata"](
+        SessionId.unwrap(cmd.sessionId),
+        metadata
+      )
+      
+      // Initialize lastServerRequestId for this session
+      val stateWithRequestId = stateWithMetadata.updated["lastServerRequestId"](
+        SessionId.unwrap(cmd.sessionId),
+        RequestId(0)
+      )
+      
+      // Call user's session created handler (passing capabilities)
+      val (newState, serverRequests) = handleSessionCreated(cmd.sessionId, cmd.capabilities).run(stateWithRequestId)
+      
+      // Add any server requests
+      val (finalState, assignedRequests) = addServerRequests(
+        cmd.createdAt,
+        newState,
+        cmd.sessionId,
+        serverRequests
+      )
+      
+      (finalState, assignedRequests)
+    }
+  
+  /**
+   * Handle SessionExpired command (internal name to avoid conflict with abstract method).
+   */
+  private def handleSessionExpired_internal(cmd: SessionCommand.SessionExpired[SR]): State[HMap[CombinedSchema[UserSchema]], List[SR]] =
+    State.modify { state =>
+      // Get capabilities from metadata before expiring
+      val capabilities = state.get["metadata"](SessionId.unwrap(cmd.sessionId))
+        .map(_.capabilities)
+        .getOrElse(Map.empty[String, String])
+      
+      // Call user's session expired handler with capabilities
+      val (newState, serverRequests) = handleSessionExpired(cmd.sessionId, capabilities).run(state)
+      
+      // Remove all session data
+      val finalState = expireSession(newState, cmd.sessionId)
+      
+      // Return any final server requests (but don't add them to pending - session is expired)
+      (finalState, serverRequests)
+    }
+  
+  /**
+   * Handle GetRequestsForRetry command.
+   * 
+   * NOTE: Left unimplemented - requires access to HMap internal iteration
+   * which is not currently available. This functionality should be implemented
+   * externally or requires additional HMap methods.
+   */
+  /**
+   * Handle GetRequestsForRetry - atomically get eligible requests and update lastSentAt.
+   */
+  private def handleGetRequestsForRetry(cmd: SessionCommand.GetRequestsForRetry[SR]): State[HMap[CombinedSchema[UserSchema]], List[PendingServerRequest[SR]]] =
+    State.modify { state =>
+      // Get pending requests for this session
+      val pending = state.get["serverRequests"](SessionId.unwrap(cmd.sessionId))
+        .getOrElse(List.empty[PendingServerRequest[SR]])
+      
+      // Filter requests eligible for retry (lastSentAt < lastSentBefore)
+      val eligible = pending.filter(req => req.lastSentAt.isBefore(cmd.lastSentBefore))
+      
+      // Update lastSentAt to createdAt (current time) for eligible requests
+      val updated = eligible.map(_.copy(lastSentAt = cmd.createdAt))
```

Comment:
this code is ugly, you overthinking. Go over the entire pending as well, do map on it, then run an if on the last sent at, update some, leave as is the rest.